controllers/configuration/backend/s3.go

```diff
@@ -0,0 +1,190 @@
+	s3AccessKey    = "AWS_ACCESS_KEY_ID"
+	s3SecretKey    = "AWS_SECRET_ACCESS_KEY"
+	s3SessionToken = "AWS_SESSION_TOKEN"
+)
+
+// S3Backend is used to interact with the Terraform s3 backend
+type S3Backend struct {
+	client    s3iface.S3API
+	AccessKey string
+	SecretKey string
+	Token     string
+	Region    string
+	Key       string
+	Bucket    string
+}
+
+func newS3BackendFromInline(ctx k8sContext, backendConfig *ParsedBackendConfig, optionSource *OptionSource) (Backend, error) {
+	region, err := backendConfig.getAttrString("region")
+	if err != nil {
+		return nil, err
+	}
+	key, err := backendConfig.getAttrString("key")
+	if err != nil {
+		return nil, err
+	}
+	bucket, err := backendConfig.getAttrString("bucket")
+	if err != nil {
+		return nil, err
+	}
+	s3Backend := &S3Backend{
+		Region: region,
+		Key:    key,
+		Bucket: bucket,
+	}
+	if err := s3Backend.fillOptions(ctx, optionSource); err != nil {
+		return nil, err
+	}
+	if err := s3Backend.buildClient(); err != nil {
+		return nil, err
+	}
+	return s3Backend, nil
+}
+
+func newS3BackendFromExplicit(ctx k8sContext, backendConfig interface{}, optionSource *OptionSource) (Backend, error) {
+	conf, ok := backendConfig.(*v1beta2.S3BackendConf)
+	if !ok || conf == nil {
+		return nil, errors.New("invalid backendConf")
+	}
+	s3Backend := &S3Backend{
+		Region: conf.Region,
+		Key:    conf.Key,
+		Bucket: conf.Bucket,
+	}
+	if err := s3Backend.fillOptions(ctx, optionSource); err != nil {
+		return nil, err
+	}
+	if err := s3Backend.buildClient(); err != nil {
+		return nil, err
+	}
+	return s3Backend, nil
+}
+
+func (s *S3Backend) fillOptions(ctx k8sContext, optionSource *OptionSource) error {
+	accessKey, ok, err := optionSource.getOption(ctx, s3AccessKey)
+	if err != nil || !ok {
+		return fmt.Errorf("get option %s error", s3AccessKey)
+	}
+	s.AccessKey = accessKey
+
+	secretKey, ok, err := optionSource.getOption(ctx, s3SecretKey)
+	if err != nil || !ok {
+		return fmt.Errorf("get option %s error", s3SecretKey)
+	}
+	s.SecretKey = secretKey
+
+	token, ok, err := optionSource.getOption(ctx, s3SessionToken)
+	if err != nil || !ok {
+		s.Token = ""
+	}
+	s.Token = token
+
+	return nil
+}
+
+func (s *S3Backend) buildClient() error {
+	sessionOpts := session.Options{
+		Config: aws.Config{
+			Credentials: credentials.NewStaticCredentials(s.AccessKey, s.SecretKey, s.Token),
+			Region:      aws.String(s.Region),
+		},
+	}
+	sess, err := session.NewSessionWithOptions(sessionOpts)
+	if err != nil {
+		return err
+	}
+	s.client = s3.New(sess)
+	return nil
+}
+
+func (s *S3Backend) getObject() (*s3.GetObjectOutput, error) {
+	input := &s3.GetObjectInput{
+		Key:    &s.Key,
+		Bucket: &s.Bucket,
+	}
+	return s.client.GetObject(input)
+}
+
+// GetTFStateJSON gets Terraform state json from the Terraform s3 backend
+func (s *S3Backend) GetTFStateJSON(_ context.Context) ([]byte, error) {
+	output, err := s.getObject()
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = output.Body.Close() }()
```

Comment:
`def output.Body.Close()`? Which is better?

Reply:
We need to handle the error returned by `output.Body.Close` explicitly, otherwise the golongci-lint will throw `errorcheck` error like this:

```
controllers/configuration/backend/s3.go:122:25: Error return value of `output.Body.Close` is not checked (errcheck)
        defer output.Body.Close()
                               ^
```

Reply:
Good, that's a good solution or you can comment to ignore the warning.